ikarus/io/vtkwriter.hh

```diff
@@ -0,0 +1,182 @@
+        assembler_(assembler) {}
+
+  template <class... Args>
+  Writer(std::shared_ptr<Assembler> assembler, DataCollector& dc, Args... args)
+      : writer_(dc, args...),
+        assembler_(assembler) {}
+
+  template <class... Args>
+  Writer(std::shared_ptr<Assembler> assembler, DataCollector&& dc, Args... args)
+      : writer_(std::move(dc), args...),
+        assembler_(assembler) {}
+
+  // Sets the VTK file format
+  void setFormat(Dune::Vtk::FormatTypes format) { writer_.setFormat(format); }
+
+  /// Sets the global datatype used for coordinates and other global float values
+  void setDatatype(Dune::Vtk::DataTypes datatype) { writer_.setDatatype(datatype); }
+
+  /// Sets the integer type used in binary data headers
+  void setHeadertype(Dune::Vtk::DataTypes datatype) { writer_.setHeadertype(datatype); }
+
+  auto& writer() const { return writer_; }
+  auto& writer() { return writer_; }
+
+  template <typename RF, Impl::Concepts::DataType DT>
+  void addResultFunction(RF&& resultFunction, DT /*type */) {
+    if constexpr (std::is_same_v<DT, Impl::AsCellData>)
+      writer_.addCellData(std::forward<RF>(resultFunction));
+    else
+      writer_.addPointData(std::forward<RF>(resultFunction));
+  }
+
+  template <template <typename, int, int> class RT, Impl::Concepts::DataType DT>
+  requires(Concepts::ResultType<RT>)
+  void addResult(DT type) {
+    auto resFunction = makeResultVtkFunction<RT>(assembler_);
+    addResultFunction(std::move(resFunction), type);
+  }
+
+  template <Impl::Concepts::DataType DT>
+  void addAllResults(DT type) {
+    using ResultTuple = typename FEType::SupportedResultTypes;
+
+    Dune::Hybrid::forEach(ResultTuple(), [&]<typename RT>(RT i) { addResult<RT::template Rebind>(type); });
+  }
+
+  template <typename GF, Impl::Concepts::DataType DT>
+  void addGridFunction(GF&& gridFunction, const std::string& name, size_t size, DT /* type */) {
+    auto fieldInfo = Dune::Vtk::FieldInfo(name, size);
+    if constexpr (std::is_same_v<DT, Impl::AsCellData>)
+      writer_.addCellData(std::forward<GF>(gridFunction), fieldInfo);
+    else
+      writer_.addPointData(std::forward<GF>(gridFunction), fieldInfo);
+  }
+
+  template <typename GF, typename FieldInfo, Impl::Concepts::DataType DT>
+  requires(std::same_as<FieldInfo, Dune::VTK::FieldInfo> or std::same_as<FieldInfo, Dune::Vtk::FieldInfo>)
+  void addGridFunction(GF&& gridFunction, FieldInfo fieldInfo, DT /* type */) {
+    if constexpr (std::is_same_v<DT, Impl::AsCellData>)
+      writer_.addCellData(std::forward<GF>(gridFunction), fieldInfo);
+    else
+      writer_.addPointData(std::forward<GF>(gridFunction), fieldInfo);
+  }
+
+  template <int dim, typename R, typename Basis, Impl::Concepts::DataType DT>
+  void addInterpolation(R&& vals, const Basis& basis, const std::string& name, DT type) {
+    auto gridFunction = Dune::Functions::makeDiscreteGlobalBasisFunction<Dune::FieldVector<double, dim>>(basis, std::forward<R>(vals));
```

Comment:
Instead of `Dune::FieldVector`, may be the type should be templated such that it also works for a particular `subspaceBasis`?

Reply:
So in my understanding this `FieldVector` serves as a Data container per node. The function is templated with the size of said container. So this should work with any kind of subspacebasis. 
If you want to make a custom gridFunction you could still use the function `addGridFunction()`

Reply:
The tempolate of makeDiscreteGlobalBasisFunction is the requested return value ofthe created function.

The `operator(local )` the returns this type. Therefore, for 2D dispalcements the full basis should retunr FieldVector<double,2> but the subspaces for the x and y dispalcements sohuld be return <double,1> if I m not mistaken.

Reply:
The signature basically should be:


```cpp
  template <typename ReturnType, Impl::Concepts::DataType DT, typename... Args>
  void addInterpolation(R&& vals, const Basis& basis, const std::string& name, DT type) {
    auto gridFunction = Dune::Functions::makeDiscreteGlobalBasisFunction<ReturnType>(basis, std::forward<R>(vals));

```

Reply:
Ive added the option to define your own Container, but it defaults to a Dune::FieldVector of a size deduced by the basis